db/types.go

```diff
@@ -0,0 +1,196 @@
+	// CONTRACT: key readonly []byte
+	Delete([]byte) error
+
+	// Flushes pending writes and discards the transaction.
+	// TODO: maybe change to Flush() and follow WriteBatch semantics (ie. don't discard)
+	Commit() error
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// DBReadWriter is a transaction interface that allows both reading and writing.
+type DBReadWriter interface {
+	DBReader
+	DBWriter
+}
+
+// Iterator represents an iterator over a domain of keys. Callers must call Close when done.
+// No writes can happen to a domain while there exists an iterator over it, some backends may take
+// out database locks to ensure this will not happen.
+//
+// Callers must make sure the iterator is valid before calling any methods on it, otherwise
+// these methods will panic. This is in part caused by most backend databases using this convention.
+//
+// As with DBReader, keys and values should be considered read-only, and must be copied before they are
+// modified.
+//
+// Typical usage:
+//
+// var itr Iterator = ...
+// defer itr.Close()
+//
+// for ; itr.Valid(); itr.Next() {
+//   k, v := itr.Key(); itr.Value()
+//   ...
+// }
+// if err := itr.Error(); err != nil {
+//   ...
+// }
+type Iterator interface {
+	// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
+	// CONTRACT: start, end readonly []byte
+	Domain() (start []byte, end []byte)
+
+	// Valid returns whether the current iterator is valid. Once invalid, the Iterator remains
+	// invalid forever.
+	Valid() bool
+
+	// Next moves the iterator to the next key in the database, as defined by order of iteration.
+	// If Valid returns false, this method will panic.
+	Next()
+
+	// Key returns the key at the current position. Panics if the iterator is invalid.
+	// CONTRACT: key readonly []byte
+	Key() (key []byte)
+
+	// Value returns the value at the current position. Panics if the iterator is invalid.
+	// CONTRACT: value readonly []byte
+	Value() (value []byte)
+
+	// Error returns the last error encountered by the iterator, if any.
+	Error() error
+
+	// Close closes the iterator, relasing any allocated resources.
+	Close() error
+}
+
+// VersionSet specifies a set of existing versions
+type VersionSet interface {
+	// Last returns the most recent saved version, or 0 if none.
+	Last() uint64
+	// Count returns the number of saved versions.
+	Count() int
+	// Iterator returns an iterator over all saved versions.
+	Iterator() VersionIterator
+	// Equal returns true iff this set is identical to another.
+	Equal(VersionSet) bool
+	// Exists returns true if a saved version exists.
+	Exists(uint64) bool
+}
+
+type VersionIterator interface {
+	// Next advances the iterator to the next element.
+	Next()
+	// Value returns the version ID at the current position.
+	Value() uint64
+	// Valid returns whether the iterator is valid. Once invalid, the iterator remains invalid.
+	Valid() bool
```

Comment:
As long as `VersionSet` is a simple self-contained object, I think it's safe to omit an `Error` method. Otherwise we can add one, but even so I think an error from `Next` is redundant.

Reply:
So I think Valid is redundant. We can have `Next()` returning a bool and `Error()` returning an error. This is the design of [sql](https://pkg.go.dev/database/sql#Rows.Next) package from the standard library. IMHO we should follow it and keep the interface minimal.